### PR TITLE
feat(rewards): run sync custom rewards concurrently

### LINKED ISSUE #129: [Task 19]  Concurrent Execution for Custom Rewards
Task 19

## 1. Repository, Baseline Commit, and Development Branch

- Repository: https://github.com/redai-infra/Relax
- Baseline branch: `main`
- Baseline commit: `039ce876d25540adad847d4223b4de4722d8f425`
- Development branch: `feat/task19-concurrent-custom-rewards`

Development and testing will be based on the commit above. The Pull Request will not contain changes unrelated to Task 19.

## 2. Runtime Environment

| Item         | Configuration                                                |
| ------------ | ------------------------------------------------------------ |
| Image        | `ghcr.io/redai-infra/relaxrl:dev-20260715-8325919e`          |
| Image digest | `sha256:3fa8ce578acda6c829b83016bde42c38fa892681e4f36ca330f545616fe578e2` |
| Python       | 3.12.3                                                       |
| CUDA         | 12.9                                                         |
| PyTorch      | 2.11.0+cu129                                                 |
| Ray          | 2.56.0                                                       |
| SGLang       | 0.5.12.post1                                                 |
| GPU          | 1 × NVIDIA H800 PCIe 80 GB                                   |
| CPU          | 2 × Intel Xeon Platinum 8558, 96 cores / 192 threads         |
| Memory       | 1.0 TiB                                                      |
| OS           | Ubuntu 22.04.5 LTS                                           |

## 3. Launch Script and Complete Commands

This task does not add a training launch script or introduce or modify CLI arguments. The existing `--custom-rm-path`, `--reward-max-concurrency`, and `--reward-num-workers` arguments will remain compatible and control custom reward execution.

Start the test image:

```bash
docker run --rm -it \
  --network host \
  --shm-size=64g \
  -v /data/zn/Relax:/workspace/Relax \
  ghcr.io/redai-infra/relaxrl:dev-20260715-8325919e \
  bash
```

Run targeted Task 19 unit tests:

```bash
cd /workspace/Relax
python -m pytest -v tests/engine/rewards/test_custom_reward_executor.py
```

Run RewardWorker regression tests:

```bash
cd /workspace/Relax
python -m pytest -v \
  tests/engine/rewards/test_reward_worker.py \
  tests/engine/rewards/test_custom_reward_executor.py
```

Run the agentic reward path regression tests:

```bash
cd /workspace/Relax
python -m pytest -v tests/test_agentic_rollout.py
```

Run the reward regression suite:

```bash
cd /workspace/Relax
python -m pytest -v tests/engine/rewards/
```

Run code checks before submission:

```bash
cd /workspace/Relax
pre-commit run --all-files
```

- New parameters: none
- Modified parameters: none

### Related Files

The main files are expected to be:

- `relax/engine/rewards/__init__.py`
- `relax/agentic/pipeline/reward.py`
- `tests/engine/rewards/test_custom_reward_executor.py`
- `tests/engine/rewards/task19_custom_rewards.py`

### Core Functions

The planned execution strategy is:

```text
--custom-rm-path
→ load and cache the custom reward function
→ determine whether the function is synchronous or asynchronous
→ directly await async functions in the caller process
→ dispatch sync functions to the RewardWorker process pool
→ enforce global concurrency through reward_max_concurrency
→ control worker process count through reward_num_workers
→ return the reward or a sample-aware exception
```

The planned changes are:

1. Cache custom reward functions by `custom_rm_path` in `RewardExecutor` to avoid loading them for every sample.
2. Detect whether a custom reward function is synchronous or asynchronous.
3. Preserve async custom reward behavior by directly awaiting it in the caller process without creating a RewardWorker.
4. Dispatch sync custom rewards to the existing RewardWorker pool in independent Ray Actor processes so they do not block the rollout event loop.
5. Cache custom reward functions by path inside each RewardWorker so each function is loaded once per process instead of once per sample.
6. Continue using the global semaphore controlled by `reward_max_concurrency`.
7. Use `reward_num_workers` to control the number of RewardWorker processes available to sync custom rewards.
8. Execute non-group custom reward batches concurrently per sample while preserving input result order.
9. Preserve whole-group custom reward behavior when `group_rm=True`.
10. Include the custom reward path and sample `group_index`, `index`, `session_id`, or batch position when loading or execution fails.
11. Ensure exceptions and cancellations release concurrency permits so later reward requests can still complete.
12. Make agentic reward paths delegate to `relax.engine.rewards`, preventing them from bypassing RewardExecutor and the Worker pool.

### Non-goals

This task will not:

- modify the Task 18 reward registry or format-aware router;
- modify GRPO, PPO, DAPO, or other RL algorithms;
- modify Controller, Ray Serve deployment, or rollout request scheduling;
- modify the public `Sample` data structure;
- add third-party dependencies;
- change existing built-in reward formulas;
- change training dataset formats;
- delete or rename existing public APIs;
- change the `--custom-rm-path` function signature;
- optimize GPU performance, training throughput, or convergence;
- modify CUDA, drivers, or the distributed training environment;
- force async custom rewards into Worker processes.

### Baseline Data and Behavior

At commit `039ce876d25540adad847d4223b4de4722d8f425`:

- `--custom-rm-path` supports loading a custom reward function;
- custom rewards are called in the rollout event-loop process;
- async custom rewards can be directly awaited;
- sync custom rewards do not reuse RewardWorker process isolation or concurrency controls;
- custom reward functions are repeatedly passed through `load_function` in different execution paths;
- `batched_async_rm` passes the complete sample list to a custom reward in one call;
- the agentic reward path bypasses `RewardExecutor` and loads and calls custom rewards directly;
- custom reward exceptions do not consistently include sample location context.

Unit tests will use minimal in-memory `Sample` objects without downloading models or training datasets and without using a GPU. Coverage will include:

- sync custom reward execution in independent processes;
- direct awaiting of async custom rewards;
- the `reward_max_concurrency` concurrency limit;
- the `reward_num_workers` Worker limit;
- custom reward caching and one-time loading;
- per-sample execution and result ordering for non-group batches;
- sync and async group reward compatibility;
- failing-sample identification;
- semaphore release and subsequent execution after an exception;
- unified agentic reward integration.

### Acceptance Metrics

| Requirement               | Measurement                                                  | Target                                                       |
| ------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
| Sync process isolation    | Compare the PID returned by the custom reward with the caller PID | PIDs are different                                           |
| Async compatibility       | Compare the async custom reward PID and check Worker creation | PID is unchanged and no RewardWorker is created              |
| Maximum concurrency       | Calculate peak overlapping intervals from sync reward start/end times | Peak concurrency does not exceed `reward_max_concurrency`     |
| Worker count              | Count unique PIDs returned by enough concurrent tasks        | Process count equals the configured `reward_num_workers`     |
| Function loading          | Record function load counts in the caller and each Worker process | Once per path in each process                                 |
| Per-sample batch execution | Execute sync custom rewards for distinct samples             | Every sample is executed and result order matches input order |
| Group compatibility       | Run sync and async group custom rewards                      | The complete group is called once and returns its reward list |
| Error identification      | Trigger an exception for a specified sample                  | Error includes the path and sample index or batch position   |
| Error release             | With concurrency set to 1, fail one sample and then run a valid sample | The next sample completes within the timeout without deadlock |
| Agentic integration       | Inspect the actual delegate used by agentic helpers          | Delegates to `relax.engine.rewards`                          |
| Reward regression         | `tests/engine/rewards/`                                      | All pass                                                     |
| Agentic regression        | `tests/test_agentic_rollout.py`                              | All pass                                                     |
| Code checks               | `pre-commit run --all-files`                                 | All pass                                                     |

Tests will use deterministic inputs without random sampling. The primary acceptance criterion is that the targeted Task 19 unit tests, reward regression suite, agentic regression tests, and code checks pass completely.

## 6. Expected Delivery Method
Delivery method: GitHub Issue + Pull Request.

### PR DESCRIPTION
**Pull request by aayc123**
# ⭐ Task19

## Execute Synchronous Custom Rewards in Worker Processes

- Detect synchronous and asynchronous custom reward callables
- Dispatch synchronous callables to the existing Ray RewardWorker pool
- Cache reward functions in the caller and each Worker process
- Preserve direct awaiting for async rewards and whole-group invocation for group rewards
- Attach reward path and sample context to custom reward failures
- Route agentic reward helpers through the shared RewardExecutor
- Recover Worker capacity when cancellation leaves a synchronous Actor method blocked

---

# ✅ Tests

## Cover Concurrent Execution and Compatibility

- Verify process isolation, Worker count, concurrency limits, and result ordering
- Verify function caching, hot reload, and async execution in the caller process
- Verify legacy batch compatibility and explicit per-sample opt-in behavior
- Verify group, custom-bypass, empty-batch, failure, and cancellation semantics
- Verify cancellation recovery with a real Ray Actor
- Verify agentic reward helpers use the shared Executor

## What

This PR adds process-pool execution for synchronous functions configured through `--custom-rm-path`:

- Synchronous custom rewards run in independent RewardWorker processes
- Asynchronous custom rewards continue to be awaited in the caller process
- Existing batched custom rewards keep receiving `rm_function(args, samples)` by default
- `--custom-rm-per-sample` explicitly opts non-group batches into concurrent per-sample execution
- `group_rm=True` always preserves whole-group invocation
- Agentic and standard rollout paths share the same RewardExecutor

## Why

Closes #129

Previously, custom rewards were called inside the rollout event loop and uniformly awaited. Synchronous functions could block the event loop and did not reuse the existing RewardWorker process isolation, Worker count, or concurrency controls. The agentic reward path also bypassed RewardExecutor, resulting in inconsistent behavior across rollout modes.

## How

- Cache custom reward callables by path and reload token in RewardExecutor
- Detect whether each callable is synchronous or asynchronous
- Directly await async callables within the existing semaphore
- Dispatch sync callables to RewardWorkers using round-robin selection
- Cache functions by path inside each Worker to avoid per-sample loading
- Enforce global concurrency through `reward_max_concurrency`
- Control Worker process count through `reward_num_workers`
- Preserve legacy batch invocation unless `--custom-rm-per-sample` is explicitly set
- Include the custom reward path and sample index, session ID, or batch position in errors
- Cancel and drain sibling tasks after failure or cancellation
- Probe a cancelled Worker and kill/recreate it when a running synchronous method remains blocked

## Compatibility and Migration

- `--custom-rm-per-sample` is a new opt-in flag and defaults to disabled
- Existing batch custom rewards require no changes and keep `def custom_rm(args, samples)`
- To opt into concurrent per-sample execution, change the second parameter to `sample` and add `--custom-rm-per-sample`
- `group_rm=True` takes precedence and always passes the complete sample list once
- Async custom rewards and built-in reward calculations remain compatible
- English and Chinese migration guidance is included in the configuration documentation

## Review Fixes

- Replace a sync RewardWorker after best-effort cancellation cannot free it, verified with a real Ray Actor
- Rename `task19_custom_rewards.py` to `custom_reward_fixtures.py`
- Preserve the legacy batch custom-reward contract by default and add an explicit compatibility switch

## Testing

Environment:

- Ray 2.56.0
- CPU-only Ray tests; no GPU required

Command:

```bash
python -m pytest -q \
  tests/engine/rewards/test_reward_worker.py \
  tests/engine/rewards/test_custom_reward_executor.py \
  tests/test_agentic_rollout.py
```

Result:

```text
80 passed, 16 warnings
```

All warnings are pre-existing Ray Serve deprecation warnings for legacy Pydantic APIs and are unrelated to this PR.

Code checks:

```bash
pre-commit run --all-files --show-diff-on-failure
```

Result: all checks passed.

- [x] `pre-commit run --all-files` passes
- [x] Targeted and related regression tests pass (80 passed)
- [ ] Full `pytest tests/` was not run because it includes GPU and distributed tests outside this task's scope
- [x] New tests added
- [x] English and Chinese migration documentation added

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] CI/CD or build changes

### CODE REVIEW

File: relax/engine/rewards/__init__.py [hunk 1]
```diff
@@ -147,6 +185,70 @@ def _next_worker(self):
...
+
+        return is_async(function) or is_async(getattr(function, "__call__", None))
+
+    @staticmethod
+    async def _await_actor_ref(ref):
+        try:
+            return await ref
+        except asyncio.CancelledError:
+            try:
+                ray.cancel(ref, force=False, recursive=True)
```
Comment (NINGBENZHE) [inline relax/engine/rewards/__init__.py]: 对正在运行的同步 Actor 方法，ray.cancel(..., force=False) 只是 best-effort，custom reward 若不主动检查取消状态，worker 仍可能被长期占用。建议明确该限制，或增加超时后 kill/recreate worker 的恢复机制，并用真实 Ray Actor 验证取消后下一请求仍可完成。

File: tests/engine/rewards/custom_reward_fixtures.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
+# Copyright (c) 2026 Relax Authors. All Rights Reserved.
```
Comment (NINGBENZHE) [inline tests/engine/rewards/custom_reward_fixtures.py:1]: 文件名不要出现task

File: relax/engine/rewards/__init__.py [hunk 3]
```diff
@@ -273,11 +379,28 @@ async def batched_async_rm(
...
-        rm_function = load_function(args.custom_rm_path)
-        return await rm_function(args, samples, **kwargs)
-    tasks = [async_rm(args, sample, **kwargs) for sample in samples]
-    rewards = await asyncio.gather(*tasks)
-    return rewards
+) -> list[int | float | dict[str, Any]]:
+    use_custom_reward = args.custom_rm_path is not None and not kwargs.get("ignore_custom", False)
+    if use_custom_reward and getattr(args, "group_rm", False):
+        # Group reward functions receive the complete group in one call.
+        return await async_rm(args, samples, **kwargs)
```
Comment (NINGBENZHE) [inline relax/engine/rewards/__init__.py:437]: 这里将 group_rm=False 的 custom reward 从原来的 rm_function(args, samples) 改成了逐样本调用 rm_function(args, sample)，会静默改变第二个参数类型并破坏现有 batch custom reward。建议保留旧批量契约，或增加显式模式/兼容开关，并补充旧接口回归测试和迁移说明。

**Review by NINGBENZHE — CHANGES_REQUESTED** [review]

**Comment by aayc123** [comment]
相对上一次提交 `d8f106c`，实际修改范围如下：

| 文件                                                         | 目的                                                         |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| relax/engine/rewards/__init__.py| 增加取消后的 Worker 探活、kill/recreate、逐 Worker 锁；实现 legacy batch 与显式逐样本模式的路由 |
| relax/utils/arguments.py | 新增默认关闭的 `--custom-rm-per-sample` 兼容开关             |
| tests/engine/rewards/test_custom_reward_executor.py | 补充旧接口、opt-in 并发、顺序、异常、空 batch、取消恢复等测试 |
| `tests/engine/rewards/task19_custom_rewards.py`              | 删除旧文件名                                                 |
| tests/engine/rewards/custom_reward_fixtures.py  | 上述文件的新名称，并增加真实 Actor 取消测试 fixture          |
| docs/en/guide/configuration.md  | 增加兼容开关和迁移说明                                       |
| docs/zh/guide/configuration.md | 同步中文迁移说明                                             |

三个 review 问题均已解决：

1. `ray.cancel(force=False)` 无法保证停止同步 Actor

   已增加取消后探活；1 秒内 Worker 仍不可用就 kill/recreate。真实 Ray Actor 测试验证了 30 秒阻塞任务取消后，下一请求能在 10 秒内由新 Worker 完成。

2. 文件名不能包含 `task`

   已将 `task19_custom_rewards.py` 重命名为 `custom_reward_fixtures.py`，代码和文档中没有残留 `task19`/`Task 19`。

3. `group_rm=False` 静默破坏旧 batch 接口

   已改为默认保持 `rm_function(args, samples)`。只有显式设置 `--custom-rm-per-sample` 才调用 `rm_function(args, sample)`；同时补充旧接口回归测试和中英文迁移说明。

   行为如下：

   - 默认不设置 --custom-rm-per-sample
     - 保持旧契约 `rm_function(args, samples)`。
     - 现有 custom reward 无需修改。
   - 显式设置 --custom-rm-per-sample
     - `group_rm=False` 时逐样本调用 `rm_function(args, sample)`。
     - 同步函数通过多个 RewardWorker 并发执行。
     - 结果保持输入顺序，异常包含 reward 路径和 batch position。
   - `group_rm=True` 始终整组调用，优先级高于新开关。
   - `ignore_custom=True` 始终进入原有内置 reward 路径

**Comment by NINGBENZHE** [comment]
验收意见：通过！当前实现已覆盖同步 custom reward 的进程隔离、并发参数生效、async 函数直接 await、函数缓存、异常定位及取消恢复，相关测试和 CI 均已通过，因此本任务验收通过。  
但新增了 --custom-rm-per-sample、reload token、阻塞 worker 探测和强制替换等机制，导致改动范围和维护成本更高。强制替换 worker 还可能使该 actor 上已排队的其他请求失败。  
因此本 PR 可按任务完成度验收通过。感谢贡献！

**Comment by SigureMo** [comment]
#181 merged，本 PR close～
